assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,339 @@
+
+    let mut context = ParserContext {
+        procedures: BTreeMap::new(),
+        imports,
+    };
+
+    context.parse_procedures(&mut tokens, false)?;
+
+    // make sure program body is present
+    let next_token = tokens
+        .read()
+        .ok_or_else(|| AssemblyError::unexpected_eof(tokens.pos()))?;
+    if next_token.parts()[0] != Token::BEGIN {
+        return Err(AssemblyError::unexpected_token(next_token, Token::BEGIN));
+    }
+
+    let program_start = tokens.pos();
+    // consume the 'begin' token
+    let header = tokens.read().expect("missing program header");
+    header.validate_begin()?;
+    tokens.advance();
+
+    // make sure there is something to be read
+    let start_pos = tokens.pos();
+    if tokens.eof() {
+        return Err(AssemblyError::unexpected_eof(start_pos));
+    }
+
+    let mut body = Vec::<Node>::new();
+
+    // parse the sequence of nodes and add each node to the list
+    let mut end_of_nodes = false;
+    let beginning_node_count = body.len();
+    while !end_of_nodes {
+        let node_count = body.len();
+        context.parse_body(&mut tokens, &mut body)?;
+        end_of_nodes = body.len() == node_count;
+    }
+
+    // make sure at least one block has been read
+    if body.len() == beginning_node_count {
+        let start_op = tokens.read_at(start_pos).expect("no start token");
+        return Err(AssemblyError::empty_block(start_op));
+    }
+
+    // consume the 'end' token
+    match tokens.read() {
+        None => Err(AssemblyError::unmatched_begin(
+            tokens.read_at(program_start).expect("no begin token"),
+        )),
+        Some(token) => match token.parts()[0] {
+            Token::END => token.validate_end(),
+            Token::ELSE => Err(AssemblyError::dangling_else(token)),
+            _ => Err(AssemblyError::unmatched_begin(
+                tokens.read_at(program_start).expect("no begin token"),
+            )),
+        },
+    }?;
+    tokens.advance();
+
+    // make sure there are no instructions after the end
+    if let Some(token) = tokens.read() {
+        return Err(AssemblyError::dangling_ops_after_program(token));
+    }
+
+    let program = ProgramAst {
+        body,
+        procedures: context.procedures,
+    };
+
+    Ok(program)
+}
+
+pub fn parse_module(source: &str) -> Result<ModuleAst, AssemblyError> {
+    let mut tokens = TokenStream::new(source)?;
+
+    let imports = parse_imports(&mut tokens)?;
+    let mut context = ParserContext {
+        procedures: BTreeMap::new(),
+        imports,
+    };
+    context.parse_procedures(&mut tokens, true)?;
+
+    // make sure program body is absent and no more instructions.
+    if tokens.read().is_some() {
+        return Err(AssemblyError::unexpected_eof(tokens.pos()));
+    }
+
+    let module = ModuleAst {
+        procedures: context.procedures,
+    };
+
+    Ok(module)
+}
+
+fn parse_op_token(op: &Token) -> Result<Node, AssemblyError> {
```

Comment:
This is called only from methods of `ParserContext`, right? If so, let's move this function into that module.

Reply:
fixed